### PR TITLE
[SPARK-38768][SQL] Remove `Limit` from plan if complete push down limit to data source.

### PR DESCRIPTION
**Pull request by beliefer**
### What changes were proposed in this pull request?
Currently, Spark supports push down limit to data source.
If limit could pushed down and Data source only have one partition, DS V2 still do limit again.
This PR want remove `Limit` from plan if complete push down limit to data source.

### Why are the changes needed?
Improve performance.

### Does this PR introduce _any_ user-facing change?
'No'.
New feature.

### How was this patch tested?
Tests updated.

### CODE REVIEW

**Comment by beliefer** [comment]
ping @huaxingao cc @cloud-fan

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala [hunk 1]
```diff
@@ -380,27 +380,32 @@ object V2ScanRelationPushDown extends Rule[LogicalPlan] with PredicateHelper wit
...
-      val newChild = pushDownLimit(child, limitValue)
-      val newLocalLimit = globalLimit.child.asInstanceOf[LocalLimit].withNewChildren(Seq(newChild))
-      globalLimit.withNewChildren(Seq(newLocalLimit))
+      val (newChild, isPartiallyPushed) = pushDownLimit(child, limitValue)
+      if (isPartiallyPushed) {
+        val newLocalLimit =
+          globalLimit.child.asInstanceOf[LocalLimit].withNewChildren(Seq(newChild))
+        globalLimit.withNewChildren(Seq(newLocalLimit))
+      } else {
+        newChild
```
Comment (huaxingao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala:407]: I think there is a problem here. If `isPartiallyPushed` is false, it is assumed that `Limit` is completely pushed down so Spark doesn't do `Limit` any more. However, the `isPartiallyPushed`  false could come from the default case in `PushDownUtils`.`pushLimit`

```
  def pushLimit(scanBuilder: ScanBuilder, limit: Int): (Boolean, Boolean) = {
    scanBuilder match {
      case s: SupportsPushDownLimit if s.pushLimit(limit) =>
        (true, s.isPartiallyPushed)
      case _ => (false, false)
    }
  }
```
In this case, the `Limit` at Spark is removed wrongly.

Comment (beliefer) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/V2ScanRelationPushDown.scala]: Thank you for the reminder.

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Comment by beliefer** [comment]
@huaxingao @cloud-fan Thank you for the review.

**Pull request closed without merging**